base/src/types.rs

```diff
@@ -498,7 +496,9 @@ impl<Id, T> Type<Id, T>
     pub fn function(args: Vec<T>, ret: T) -> T {
         args.into_iter()
             .rev()
-            .fold(ret, |body, arg| T::from(Type::Function(vec![arg], body)))
+            .fold(ret, |body, arg| {
+                Type::app(Type::builtin(BuiltinType::Function), vec![arg, body])
```

Comment:
It would be a bit more efficient to lift `Type::builtin(BuiltinType::Function)` out of the fold and `.clone()` it. (#41 is a bit more relevant after removing this change as well).